trin-core/src/portalnet/overlay.rs

```diff
@@ -175,12 +190,182 @@ impl<TContentKey: OverlayContentKey + Send, TMetric: Metric + Send>
+                            self.store_accepted_content(&key, content_value.clone());
+                            content_keys_values.push((key, content_value))
+                        }
+                        Err(err) => {
+                            warn!("Unable to decode state content key: {err}");
+                            continue;
+                        }
+                    }
+                }
+                // Propagate gossip accepted content
+                self.propagate_gossip(content_keys_values);
+            }
+            _ => return Err(anyhow!("Unsupported content key protocol.")),
+        };
+        Ok(())
+    }
+
+    /// Propagate gossip accepted content via OFFER/ACCEPT:
+    fn propagate_gossip(&self, accepted_content: Vec<(impl OverlayContentKey, ByteList)>) {
+        // Get all nodes from overlay routing table
+        let kbuckets = self.kbuckets.read();
+        let all_nodes: Vec<&kbucket::Node<NodeId, Node>> = kbuckets
+            .buckets_iter()
+            .map(|kbucket| {
+                kbucket
+                    .iter()
+                    .collect::<Vec<&kbucket::Node<NodeId, Node>>>()
+            })
+            .flatten()
+            .collect();
+
+        // HashMap to temporally store all interested nodes and the content
+        let mut nodes_and_content: HashMap<String, Vec<RawContentKey>> = HashMap::new();
+
+        // Filter all nodes from overlay routing table where XOR_distance(content_id, nodeId) < node radius
+        for content in accepted_content {
+            let interested_nodes: Vec<Enr> = all_nodes
+                .clone()
+                .into_iter()
+                .filter(|node| {
+                    XorMetric::distance(&content.0.content_id(), &node.key.preimage().raw())
+                        < node.value.data_radius()
+                })
+                .map(|node| node.value.enr())
+                .collect();
+
+            // Continue if no nodes are interested in the content
+            if interested_nodes.is_empty() {
+                debug!("No nodes interested in neighborhood gossip: content_key={} num_nodes_checked={}", hex_encode(content.0.into()), all_nodes.len());
+                continue;
+            }
+
+            // Get log2 random enrs to gossip
+            let random_nodes = log2_random_enrs(interested_nodes);
+
+            // Temporally Store all random selected nodes with interested content.
+            // We want this so we can offer all the content to interested node in one request.
+            for enr in random_nodes {
+                nodes_and_content
+                    .entry(enr.to_base64())
+                    .or_default()
+                    .push(content.clone().0.into());
+            }
+        }
+
+        // Create and send OFFER overlay request to the interested nodes
+        for (nodes, interested_content) in nodes_and_content.drain() {
+            let enr = match Enr::from_str(&nodes) {
```

Comment:
I take it this conversion is because an `Enr` instance is not hashable for the `HashMap`? Otherwise would be nice to avoid these steps of `.to_base64()` and `::from_str()`.

Reply:
Yeah, it is because `discv5::enr::Enr<CombinedKey>` doesn't implement the `Hash` trait. The first workaround that came to my mind was to convert it to String.